Sources/FluentKit/Properties/Timestamp.swift

```diff
@@ -130,3 +156,103 @@ extension Schema {
         query.filters.append(.group([isNull, isFuture], .or))
     }
 }
+
+
+// MARK: - Timestamp Formatter
+
+public protocol AnyTimestampFormatter {
+    var timestampType: Codable.Type { get }
+
+    func timestamp(from date: Date) -> Codable
+    func date(from timestamp: Codable) -> Date?
+
+    func decode(at key: FieldKey, from output: DatabaseOutput) throws -> Date?
+}
+
+public protocol TimestampFormatter: AnyTimestampFormatter {
+    associatedtype Timestamp: Codable
+
+    func timestamp(from date: Date) -> Timestamp
+    func date(from timestamp: Timestamp) -> Date?
+}
+
+extension TimestampFormatter {
+    public var timestampType: Codable.Type { Timestamp.self }
+
+
+    public func timestamp(from date: Date) -> Codable {
+        return self.timestamp(from: date) as Timestamp
+    }
+
+    public func date(from timestamp: Codable) -> Date? {
+        return (timestamp as? Timestamp).flatMap(self.date(from:))
+    }
+
+
+    public func decode(at key: FieldKey, from output: DatabaseOutput) throws -> Date? {
+        let timestamp = try output.decode(key, as: Timestamp?.self)
+        return timestamp.flatMap(self.date(from:))
+    }
+}
+
+
+extension DateFormatter: TimestampFormatter {
+    public func timestamp(from date: Date) -> String { self.string(from: date) }
+}
+
+extension ISO8601DateFormatter: TimestampFormatter {
+    public func timestamp(from date: Date) -> String { self.string(from: date) }
+}
+
+public struct UnixTimestampFormatter: TimestampFormatter {
+    public func timestamp(from date: Date) -> Double { date.timeIntervalSince1970 }
+    public func date(from timestamp: Double) -> Date? { Date(timeIntervalSince1970: timestamp) }
+}
+
+public struct DefaultTimestampFormatter: TimestampFormatter {
+    public func timestamp(from date: Date) -> Date { date }
+    public func date(from timestamp: Date) -> Date? { timestamp }
+}
+
+
+private final class TimestampFormatterCache {
+    static func formatter<Formatter>(for id: String, factory: () -> Formatter) -> Formatter where Formatter: TimestampFormatter {
+        if let formatter = self.current.currentValue?.cache[id] as? Formatter { return formatter }
+
+        let new = factory()
+
+        if self.current.currentValue == nil { self.current.currentValue = TimestampFormatterCache() }
+        self.current.currentValue?.cache[id] = new
+
+        return new
+    }
+
+    private static var current: ThreadSpecificVariable<TimestampFormatterCache> = .init()
+
+    var cache: [String: AnyTimestampFormatter] = [:]
+}
+
+
+public struct TimestampFormat<Formatter> where Formatter: TimestampFormatter {
```

Comment:
What if this held `AnyTimestampFormatter` instead of a generic one? That would let us make a lot of these types private.